cmd/ipfs/ulimit_freebsd.go

```diff
@@ -3,42 +3,53 @@
 package main
 
 import (
+	"errors"
 	"fmt"
 	"syscall"
 )
 
-func init() {
-	fileDescriptorCheck = checkAndSetUlimit
-}
-
-func checkAndSetUlimit() error {
-	var rLimit syscall.Rlimit
-	err := syscall.Getrlimit(syscall.RLIMIT_NOFILE, &rLimit)
-	if err != nil {
+// checks for the maximum number of file descriptors that the system supports
+// if we have a smaller softer, maximum value than the IPFS_FD_MAX
+// current env variable then we shall overwrite them with it
+func managefd() (err error) {
+	rlimit := syscall.Rlimit{}
+	if err = syscall.Getrlimit(syscall.RLIMIT_NOFILE, &rlimit); err != nil {
 		return fmt.Errorf("error getting rlimit: %s", err)
 	}
 
-	ipfsFileDescNum := int64(ipfsFileDescNum)
-
-	var setting bool
-	if rLimit.Cur < ipfsFileDescNum {
-		if rLimit.Max < ipfsFileDescNum {
-			log.Error("adjusting max")
-			rLimit.Max = ipfsFileDescNum
+	// on freebds this is int64
+	nfds := int64(maxFds)
+
+	max := rlimit.Max
+	rlimit.Cur, rlimit.Max = nfds, nfds
+
+	// the soft limit is the value that the kernel enforces for the corresponding resource
+	// the hard limit acts as a ceiling for the soft limit
+	// an unprivileged process may only set its soft limit to a value
+	// in the range from 0 up to the hard limit
+	if err = syscall.Setrlimit(syscall.RLIMIT_NOFILE, &rlimit); err != nil {
+		// the process does not have permission so we should only set the soft value
+		if err == syscall.EPERM {
+			// if the user has passed a fds number larger than the max
+			// and the process is without
+			if rlimit.Cur > max {
+				return errors.New(
+					"Cannot set rlimit, IPFS_FD_MAX is larger than the hard limit",
+				)
+			}
+
+			rlimit.Max = max
+			// set the rlimit in unprivileged mode
+			if err = syscall.Setrlimit(syscall.RLIMIT_NOFILE, &rlimit); err != nil {
+				return fmt.Errorf("error setting ulimit without hard limit: %s", err)
+			}
+
+			log.Infof("Successfully raised file descriptor limit to %d.\n", maxFds)
```

Comment:
Info is not default on channel so it won't be visible. We want this message to be visible during daemon start, that is the reason it was using `fmt.Printf` before.